# Import libraries

In [118]:
import pandas as pd

import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots

# Import Data

In [119]:
df = pd.read_csv("src/Speed+Dating+Data.csv", encoding='ISO-8859-1')

# Basic satistics

In [120]:
print(f'The dataset contains {df.shape[0]} rows and {df.shape[1]} columns')

display(df.head())
display(df.describe(include='all'))

The dataset contains 8378 rows and 195 columns


iid   id  gender  idg  condtn  wave  round  position  positin1  order  ...  \
0    1  1.0       0    1       1     1     10         7       NaN      4  ...   
1    1  1.0       0    1       1     1     10         7       NaN      3  ...   
2    1  1.0       0    1       1     1     10         7       NaN     10  ...   
3    1  1.0       0    1       1     1     10         7       NaN      5  ...   
4    1  1.0       0    1       1     1     10         7       NaN      7  ...   

   attr3_3  sinc3_3  intel3_3  fun3_3  amb3_3  attr5_3  sinc5_3  intel5_3  \
0      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
1      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
2      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
3      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
4      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   

   fun5_3  amb5_3  
0     NaN     NaN  
1     NaN     NaN  
2     NaN     NaN  
3     NaN     NaN  
4     NaN     NaN  

[5 rows x 195 columns]

iid           id       gender          idg       condtn  \
count   8378.000000  8377.000000  8378.000000  8378.000000  8378.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean     283.675937     8.960248     0.500597    17.327166     1.828837   
std      158.583367     5.491329     0.500029    10.940735     0.376673   
min        1.000000     1.000000     0.000000     1.000000     1.000000   
25%      154.000000     4.000000     0.000000     8.000000     2.000000   
50%      281.000000     8.000000     1.000000    16.000000     2.000000   
75%      407.000000    13.000000     1.000000    26.000000     2.000000   
max      552.000000    22.000000     1.000000    44.000000     2.000000   

               wave        round     position     positin1        order  ...  \
count   8378.000000  8378.000000  8378.000000  6532.000000  8378.000000  ...   
unique          NaN          NaN          NaN          NaN          NaN  ...   
top             NaN          NaN          NaN          NaN          NaN  ...   
freq            NaN          NaN          NaN          NaN          NaN  ...   
mean      11.350919    16.872046     9.042731     9.295775     8.927668  ...   
std        5.995903     4.358458     5.514939     5.650199     5.477009  ...   
min        1.000000     5.000000     1.000000     1.000000     1.000000  ...   
25%        7.000000    14.000000     4.000000     4.000000     4.000000  ...   
50%       11.000000    18.000000     8.000000     9.000000     8.000000  ...   
75%       15.000000    20.000000    13.000000    14.000000    13.000000  ...   
max       21.000000    22.000000    22.000000    22.000000    22.000000  ...   

            attr3_3      sinc3_3     intel3_3       fun3_3       amb3_3  \
count   3974.000000  3974.000000  3974.000000  3974.000000  3974.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       7.240312     8.093357     8.388777     7.658782     7.391545   
std        1.576596     1.610309     1.459094     1.744670     1.961417   
min        2.000000     2.000000     3.000000     2.000000     1.000000   
25%        7.000000     7.000000     8.000000     7.000000     6.000000   
50%        7.000000     8.000000     8.000000     8.000000     8.000000   
75%        8.000000     9.000000     9.000000     9.000000     9.000000   
max       12.000000    12.000000    12.000000    12.000000    12.000000   

            attr5_3      sinc5_3     intel5_3       fun5_3       amb5_3  
count   2016.000000  2016.000000  2016.000000  2016.000000  2016.000000  
unique          NaN          NaN          NaN          NaN          NaN  
top             NaN          NaN          NaN          NaN          NaN  
freq            NaN          NaN          NaN          NaN          NaN  
mean       6.810020     7.615079     7.932540     7.155258     7.048611  
std        1.507341     1.504551     1.340868     1.672787     1.717988  
min        2.000000     2.000000     4.000000     1.000000     1.000000  
25%        6.000000     7.000000     7.000000     6.000000     6.000000  
50%        7.000000     8.000000     8.000000     7.000000     7.000000  
75%        8.000000     9.000000     9.000000     8.000000     8.000000  
max       10.000000    10.000000    10.000000    10.000000    10.000000  

[11 rows x 195 columns]

In [121]:
print("Missing value percentage :\n")
(df.isnull().sum()/len(df)*100)

Missing value percentage :



iid          0.000000
id           0.011936
gender       0.000000
idg          0.000000
condtn       0.000000
              ...    
attr5_3     75.936978
sinc5_3     75.936978
intel5_3    75.936978
fun5_3      75.936978
amb5_3      75.936978
Length: 195, dtype: float64

# Distributions

Participants meet 20 partners throughout the night. Therefore, there is data for each date, with 20 rows per participant.  
It is necessary to group the data to obtain accurate numbers.

## Gender

In [122]:

gender_df = df.drop_duplicates('iid')['gender']

#Maps label
gender_df=gender_df.map({0:'Female',1:'Male'})

In [123]:
fig = px.pie(
    values=gender_df.value_counts(),
    names=gender_df.unique(),
    title='Gender distribution',
    height = 600,
    width = 600
)

fig.show()

With **274 males and 277 females**, the sample is almost **perfectly balanced.**

## Age

In [124]:
age_df = df.drop_duplicates('iid')['age']

In [125]:
age_df.describe()

count    543.000000
mean      26.359116
std        3.766960
min       18.000000
25%       24.000000
50%       26.000000
75%       28.000000
max       55.000000
Name: age, dtype: float64

In [126]:
fig = px.histogram(
    age_df,
    height=600,
    width=1000
)

fig.show()

Over **543** people, **50 %** are between **24 and 28 years old** with a **mean** at **26**.  

## Race distribution

In [127]:
race_df = df.drop_duplicates('iid')['race']

In [128]:
race_df = race_df.map({1:'Black/African American',
                2:'European/Caucasian-American',
                3:'Latino/Hispanic American',
                4:'Asian/Pacific Islander/Asian-American',
                5:'Native American',
                6:'Other'})

race_df.value_counts()

race
European/Caucasian-American              304
Asian/Pacific Islander/Asian-American    136
Latino/Hispanic American                  42
Other                                     37
Black/African American                    26
Name: count, dtype: int64

In [129]:
fig = px.histogram(
    race_df,
    height=600,
    width=600
)

fig.show()

# Attributes Analysis

The documentation of the dataset explains that the method to get notes on attributes change for waves 6 to 9 with notes between 1 to 10.  
Let's drop them.

In [130]:
wave_filtered_df = df.loc[(df['wave'] < 6) | (df['wave'] > 9)]

#### Attributes descriptions :

Participants will have to fill **4** differents **survey** at different times :
1. At the event sign-up.  
2. At half-time, during the event.  
3. After a few days.
4. After a few weeks.

In each survey, they will have to note attributes :  
- **attr1_1** -> Attractive
- **sinc1_1** -> Sincere
- **intel1_1** -> Intelligent
- **fun1_1** -> Fun
- **amb1_1** -> Ambitious
- **shar1_1** -> Shared Interests

The **first number** represent the **question** type :
- **1_x** : Importance of attributes for the participant in a potential date
- **2_x** : Importance of attributes for the other sex
- **3_x** : Self-assessment
- **4_x** : Which attributes are **generally** the most important for other people.
- **5_x** : How people perceive the participant.
- **7_x** : Importance of attributes in previous decisions.

The **second number** represents the **survey** number.

- _1 : On sign-up
- _2 : After a few days
- _3 : After a few weeks
- _s : at halfway

## The attribute that people look for the least

Participants answered to 4 diffenrent survey, with the same questions regarding attributes.  
First, we will compute the mean of each attributes per participant.  
Then, we will compute mean of each attributes for throught all the dataset.  

In [ ]:
attributes_list = ['attr1','sinc1','intel1','fun1','amb1','shar1']

attributes_df = wave_filtered_df[['iid','gender']]

for attribute in attributes_list:
    #select features with substring in its name
    filtered_attribute = wave_filtered_df.filter(like=attribute)

    #compute mean of attribute as a new column
    attributes_df[f'{attribute}_mean'] = filtered_attribute.mean(axis=1)

attributes_df = attributes_df.drop_duplicates()


In [132]:
attrib_only_df = attributes_df.drop(['iid','gender'], axis=1)

mean_values = attrib_only_df.mean().sort_values(ascending=False)

highest_index = mean_values.idxmax()
lowest_index = mean_values.idxmin()

colors = ['lightblue' if (x != highest_index and x != lowest_index) 
          else 'green' if x == highest_index 
          else 'red' for x in mean_values.index]

fig = px.bar(
    mean_values,
)

fig.update_layout(
    xaxis_title="Attribute",
    yaxis_title="Mean Value",
    title="Sought-after attributes",
    showlegend=False,
    height=500,
    width=700
)

fig.update_traces(marker_color=colors)

fig.show()

mean_attributes_plot = fig

If we look at all the participants, we see that **ambition** is the **least** sought-after characteristic among partners.  
On the contrary, **attractiveness** comes first.

Let's see if the results are the same according to the participant's gender

In [133]:
by_gender_mean_values = attributes_df.drop('iid', axis=1).groupby('gender').mean()


In [134]:
by_gender_mean_values = by_gender_mean_values.reset_index()

female_values_df = by_gender_mean_values.sort_values(by=0, axis=1, ascending=False)
female_values_df= female_values_df.loc[0].drop('gender')

male_values_df = by_gender_mean_values.sort_values(by=1, axis=1, ascending=False)
male_values_df = male_values_df.loc[1].drop('gender')

In [135]:
fig = make_subplots(rows=1, cols=2, subplot_titles=("Female", "Male"), shared_yaxes=True)

fig.add_trace(go.Bar(
    x=female_values_df.index,
    y=female_values_df.values,
    name='Female',
    marker_color='lightgreen'
    ),
    row=1, col=1
    )

fig.add_trace(go.Bar(
    x=male_values_df.index,
    y=male_values_df.values,
    name='Male',
    marker_color='purple'
    ),
    row=1, col=2
)

fig.update_layout(
    title="Comparison of average attribute values by gender",
    xaxis_title="Attributes",
    yaxis_title="Mean Values",
    height=600,
    width=900
    
)



As **attractiveness** is the most sought-after characteristic for both genders, men are far more likely than women to be looking for it.

As for the least appreciated attribute, **ambition** predominates in both cases.

## Attractivness : Potential vs Real Impact

So far, we have looked at what people **think** is the most important in a partner.  
But is the impact the same in reality?


Let's look at how people who were matched by their partner have been evaluated by them

In [136]:
matched_df = wave_filtered_df[(df['dec_o'] == 1)]

C:\Users\renau\AppData\Local\Temp\ipykernel_41512\1796277692.py:1: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



In [ ]:
attributes_list = ['attr_o','sinc_o','intel_o','fun_o','amb_o','shar_o']

matched_attributes_df = matched_df[['iid','gender']]

for attribute in attributes_list:
    #select features with substring in its name
    filtered_attribute = df.filter(like=attribute)

    #compute mean of attribute as a new column
    matched_attributes_df[f'{attribute}_mean'] = filtered_attribute.mean(axis=1)

matched_attributes_df = matched_attributes_df.drop_duplicates()


In [138]:
matched_attributes_df = matched_attributes_df.drop(['iid','gender'], axis=1)

mean_values = matched_attributes_df.mean().sort_values(ascending=False)

highest_index = mean_values.idxmax()
lowest_index = mean_values.idxmin()

In [139]:
colors = ['lightblue' if (x != highest_index and x != lowest_index) 
          else 'green' if x == highest_index 
          else 'red' for x in mean_values.index]

fig = px.bar(
    x=mean_values.index,
    y=mean_values,
)

fig.update_layout(
    xaxis_title="Attribute",
    yaxis_title="Mean Value",
    title="Attributes given to matched people",
    showlegend=False,
    height=500,
    width=700
)

fig.update_traces(marker_color=colors)

fig.show()

The documentation of the dataset explains that the notation system differs for those attributes(note are from 1 to 10). 
 
Nevertheless, we can see that attributes importance are mor evenly distributed, and the first attribute is **intelligence** !  
**Attractiveness** comes fourth, and the least important is **shared interests**.

# Race analysis

In the survey, the question about the importance of race was asked.  
The distribution is uneven between races, with caucasian people being the majority.  
Nevertheless, let's take a look at how it impacts matches.

In [140]:
samerace_df = df['samerace']
samerace_df = samerace_df.map({0:'different race',1:'same race'})

In [141]:
fig = px.pie(
    values=samerace_df.value_counts(),
    names=samerace_df.unique(),
    title='Same-Race Dating Distribution',
    height = 600,
    width = 600
)

fig.show()

60,4 % of dates were between people of different race.

Among dates with partner of a different race, let's look how many get a second date.

In [142]:
#selects all dates that lead to a second one
match_df = df[(df['match'] == 1) & (df['dec_o'] == 1)]

In [143]:
total_dates = len(match_df['samerace'])
samerace_ratio = len(match_df[match_df['samerace'] == 1]) / total_dates * 100
differentrace_ratio = len(match_df[match_df['samerace'] == 0]) / total_dates * 100

In [144]:
print('"On all dates where both partners agreed to meet again :\n')
print(f'- {samerace_ratio:.2f} % were form the same race\n')
print(f'- {differentrace_ratio:.2f} % were form different race\n')

"On all dates where both partners agreed to meet again :

- 41.01 % were form the same race

- 58.99 % were form different race



We can see that participants dating people of a different race tend to match easier !

In [145]:
imprace_df = df.drop_duplicates('iid')['imprace']

In [146]:
fig = px.bar(
    imprace_df.value_counts(),
    labels={
        'value':'Count',
        'imprace':'Note',
        'variable':'Importance',
    },
    height=600,
    width=900,
    title='Importance of race for a date'
)

fig.show()

We note that a majority of people say they don't attach any importance to their partner's race.

# Conclusion

- The majority of participants are young, with an average age of 26.
- Participants don't necessarily follow their own indications as to the importance of certain attributes 
- They tend to favor partners they find **intelligent**, and pay less attention to **shared interests.** 
- Most participants don't attach importance to their partner's race.
- it would seem that meeting people of a different race increases the chances of a match.